kv/kv.go

```diff
@@ -13,98 +13,12 @@
 
 package kv
 
-import (
-	"bytes"
-	"math"
-
-	"github.com/juju/errors"
-)
-
-// Key represents high-level Key type.
-type Key []byte
-
-// Next returns the next key in byte-order.
-func (k Key) Next() Key {
-	// add 0x0 to the end of key
-	buf := make([]byte, len([]byte(k))+1)
-	copy(buf, []byte(k))
-	return buf
-}
-
-// Cmp returns the comparison result of two key.
-// The result will be 0 if a==b, -1 if a < b, and +1 if a > b.
-func (k Key) Cmp(another Key) int {
-	return bytes.Compare(k, another)
-}
-
-// EncodedKey represents encoded key in low-level storage engine.
-type EncodedKey []byte
-
-// Cmp returns the comparison result of two key.
-// The result will be 0 if a==b, -1 if a < b, and +1 if a > b.
-func (k EncodedKey) Cmp(another EncodedKey) int {
-	return bytes.Compare(k, another)
-}
-
-// Next returns the next key in byte-order.
-func (k EncodedKey) Next() EncodedKey {
-	return EncodedKey(bytes.Join([][]byte{k, Key{0}}, nil))
-}
-
-// VersionProvider provides increasing IDs.
-type VersionProvider interface {
-	CurrentVersion() (Version, error)
-}
-
-// Version is the wrapper of KV's version.
-type Version struct {
-	Ver uint64
-}
-
-var (
-	// MaxVersion is the maximum version, notice that it's not a valid version.
-	MaxVersion = Version{Ver: math.MaxUint64}
-	// MinVersion is the minimum version, it's not a valid version, too.
-	MinVersion = Version{Ver: 0}
-)
-
-// NewVersion creates a new Version struct.
-func NewVersion(v uint64) Version {
-	return Version{
-		Ver: v,
-	}
-}
-
-// Cmp returns the comparison result of two versions.
-// The result will be 0 if a==b, -1 if a < b, and +1 if a > b.
-func (v Version) Cmp(another Version) int {
-	if v.Ver > another.Ver {
-		return 1
-	} else if v.Ver < another.Ver {
-		return -1
-	}
-	return 0
-}
-
-// DecodeFn is a function that decodes data after fetching from store.
-type DecodeFn func(raw interface{}) (interface{}, error)
-
-// EncodeFn is a function that encodes data before putting into store.
-type EncodeFn func(raw interface{}) (interface{}, error)
+import "github.com/juju/errors"
 
 // ErrNotCommitted is the error returned by CommitVersion when this
 // transaction is not committed.
 var ErrNotCommitted = errors.New("this transaction is not committed")
```

Comment:
I've noticed that all other errors are listed in 'iter.go'. (not a good place either)
I think we should find a better place to declare all errors.